common/restart_waiter.cpp

```diff
@@ -0,0 +1,94 @@
+#include "restart_waiter.h"
+#include "dbconnector.h"
+#include "redispipeline.h"
+#include "select.h"
+#include "schema.h"
+#include "subscriberstatetable.h"
+#include "table.h"
+#include <string>
+
+using namespace swss;
+
+static const std::string STATE_DB_NAME = "STATE_DB";
+static const std::string STATE_DB_SEPARATOR = "|";
+static const std::string RESTART_KEY = "system";
+static const std::string RESTART_ENABLE_FIELD = "enable";
+
+bool RestartWaiter::waitRestartDone(unsigned int maxWaitSec, unsigned int db_timeout, bool isTcpConn)
+{
+    int selectTimeout = static_cast<int>(maxWaitSec);
+
+    DBConnector stateDb(STATE_DB_NAME, db_timeout, isTcpConn);
+    try
+    {
+        auto ret = stateDb.hget(STATE_WARM_RESTART_ENABLE_TABLE_NAME + STATE_DB_SEPARATOR + RESTART_KEY, RESTART_ENABLE_FIELD);
+        if (!ret || *ret.get() != "true")
+        {
+            return true;
+        }
+    }
+    catch (std::runtime_error &e)
+    {
+        return true;
+    }
+
+    SubscriberStateTable restartEnableTable(&stateDb, STATE_WARM_RESTART_ENABLE_TABLE_NAME);
+    Select s;
+    s.addSelectable(&restartEnableTable);
+
+    auto start = std::chrono::steady_clock::now();
+    while (1)
+    {
+        Selectable *sel = NULL;
+        int ret = s.select(&sel, selectTimeout * 1000, true);
+
+        if (ret == Select::OBJECT)
+        {
+            KeyOpFieldsValuesTuple kco;
+            restartEnableTable.pop(kco);
+            auto &key = kfvKey(kco);
+            if (key == RESTART_KEY)
+            {
+                auto& values = kfvFieldsValues(kco);
+                for (auto& fvt: values)
+                {
+                    auto& field = fvField(fvt);
+                    auto& value = fvValue(fvt);
+                    if (field == RESTART_ENABLE_FIELD)
+                    {
+                        if (value == "false")
+                        {
+                            return true;
+                        }
+                        else
+                        {
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        else if (ret == Select::ERROR)
+        {
+            SWSS_LOG_NOTICE("Error: wait restart done got error - %s!", strerror(errno));
+        }
+        else if (ret == Select::TIMEOUT)
+        {
+            SWSS_LOG_INFO("Timeout: wait restart done got select timeout");
+        }
+        else if (ret == Select::SIGNALINT)
+        {
+            return false;
```

Comment:
in which situation a `SIGNALINT` can be returned?

Reply:
E.g. user kills the daemon.